### ISSUE TITLE
process out of memory"

### ISSUE THREAD

**Issue by martinleopold** [simonlast/node-persist#43]

**Comment by akhoury** [comment]
doable. but needs to be added as a feature

**Comment by Vanuan** [comment]
:(
Looks like the cache is fully loaded in memory on init. Additionally, it takes much more (5x) space in memory than on the disk.

**Comment by Vanuan** [comment]
Here is it:
```javascript
            var input = self.parse(text);
...
            self.data[input.key] = input; // {key, value}
```
So instead of relying on that each filename is an md5 of key, it still reads key value from contents rather than calculating md5 on the fly. 

```javascript
    getDataValue: function (key) {
        return this.data[key] ? this.copy(this.data[key].value) : undefined;
    },
```

Much more memory efficient way would be this:

```javascript
    getDataValue: function (key) {
        return self.parseFile(md5(key)).then((data) => {
          if(data.key !== key) throw "collision detected"; 
          return data.value;
        })
    },
```

**Comment by akhoury** [comment]
^ it's not that simple, we would have to change all sync statements to async

**Comment by Vanuan** [comment]
Yeah, it would change API to be async by default. As I see, node-persist provides both sync and async versions, so that could be called getDataValueAsync and getDataValueSync.

**Comment by ErichBSchulz** [comment]
this module looked really really cool and did what it says on the box, but a full collection read on every access  is bit of a show stopper for me... I will look at rethinkdb - would be happy with a pure async library

**Comment by akhoury** [comment]
> but a full collection read on every access

that's not true, the full collection read only happens once, at init time.

**Comment by akhoury** [comment]
using this module depends on your case, if you don't want to run another process to read/write data to (`mysqld`, `mongod` etc.), then this is a good solution, but if you can, always use a full-blown well supported database instead this module.

The intent of this module was never to emulate a full blown database with a separate process running for it, like `mysql`, `mongo`, `rethinkdb` and others.. the intent of the module to emulate the [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) Web API 

from the [readme](https://github.com/simonlast/node-persist#super-easy-and-fast-persistent-data-structures-in-nodejs-modeled-after-html5-localstorage)

> .... modeled after HTML5 localStorage

Loading data in RAM comes with the cost of memory size, reading data to disk comes with the cost of the disk i/o performance, that why `redis` is fast but _possibly_ dangerous

**Comment by akhoury** [comment]
Anyways, i do see case where we should let the user decide to load everything in memory or not, the issue here is read from RAM is synchronous, reading from disk/network is not, which means if the data is never loaded in memory, we cannot provide a synchronous API.. so i guess we would always provide asynchronous and if the data is in fact in RAM then we should resolve quickly, doable but needs a major version change, since this would break previous synchronous API methods

**Comment by Vanuan** [comment]
For InMemory DB I'd look at https://github.com/Restream/reindexer. It has super fast reads.
For disk storage It uses Google's [LevelDB library](https://github.com/google/leveldb). Maybe node-persist could leverage that to solve the memory problem.
